ExternalTaskSensor respects soft_fail if the external task enters a failed_state

argibbs opened the pull request:
# Problem

You have an external task which may "skip". When that happens, you want the external task sensor to also skip. This PR enables that behaviour.

# Solution

The "right" way to do this is:
1. Set `failed_states` to `[State.SKIPPED]` (by default it is `None`)
2. Set `soft_fail=True`

With the above settings, when the target task enters the SKIPPED state, the ExternalTaskSensor will see this as a failed state and immediately react to the failure. Because of the `soft_fail` flag, it should not mark as failed, but instead as skipped.

# Effect of change

Prior to this change, the `ExternalTaskSensor` ignored the `soft_fail` setting, and always marked it as an error. With this change in place, the `ExternalTaskSensor` will instead mark the task as skipped, as desired.

# Other comments

`soft_fail` would already kick in if the ExternalTaskSensors timed out (i.e. if the external task did not enter an allowed_state or failed_state in the time. That means a possible workaround for the above is to set a timeout on the ExternalTaskSensor which you believe is long enough that the external task should have completed.

However (in my experience at least) ExternalTaskSensors are often run without timeouts. This is because if the target task hangs on upstream dependencies, having the external sensor timeout means cascading failures throughout the system which normally means more cleanup. It's usually easier to have the ExternalTaskSensors not timeout and thus only fail if there's an actual problem.

Additionally, soft_fail + execution timeouts is not perfect. If the task skips you may want the skip to propagate immediately; under the workaround, the skip would only kick in after the timeout. And of course, there are obvious goldilocks issues with finessing the timeout to be not too short nor too long.

# Related Issues

closes: #19754

argibbs committed:
Respecting soft_fail in ExternalTaskSensor when the upstream tasks are in the failed state (#19754)

- Changed behaviour of sensor to as above to respect soft_fail
- Added tests of new soft_fail behaviour (#19754)
- Added newsfragment and improved sensor docstring

potiuk reviewed (changes requested):
This looks cool, however this is a change in behaviour which should be documented. It's not a breaking change I think, I'd rather lean towards a bugfix (It should have behaved like that already). I think though this one is enough of a change to warrant a newsfragment (of bugfix type) https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#step-4-prepare-pr

Can you please add it (and also add information about this behaviour in the docstring of the External Task Sensor @argibbs ?

@jedcunningham @ephraimbuddy - this is not a "critical enough" to make into 2.3.1 so I'd leave it up to you whether to cherry-pick / add it to 2.3.1 (though it should be small and localized enought to not be a problem to cherry-pick if we alll agree this is a "bugfix".

josh-fell commented on airflow/sensors/external_task.py:
```diff
@@ -180,11 +198,20 @@ def poke(self, context, session=None):
             else:
+                if self.soft_fail:
+                    raise AirflowSkipException(
+                        f'The external DAG {self.external_dag_id} failed. ' 'Skipping due to soft_fail'
```
```suggestion
                        f'The external DAG {self.external_dag_id} failed. Skipping due to soft_fail.'
```
Cleaning up a little implicit string concatenation.

argibbs replied:
Ah sorry - this is me running black and not properly vetting the result. Will fix.

josh-fell commented on newsfragments/23647.bugfix.rst:
```diff
@@ -0,0 +1 @@
+``ExternalTaskSensor`` now supports the ``soft_fail`` flag to skip if external task or dag enters a failed state
```
```suggestion
``ExternalTaskSensor`` now supports the ``soft_fail`` flag to skip if external task or DAG enters a failed state.
```
Trivial nit.

josh-fell commented on tests/sensors/test_external_task_sensor.py:
```diff
@@ -162,10 +181,7 @@ def test_external_task_sensor_external_task_ids_param(self):
-                'dag_id': TEST_DAG_ID,
-                'task_id': TEST_TASK_ID,
-            },
+            params={'dag_id': TEST_DAG_ID, 'task_id': TEST_TASK_ID,},
```
```suggestion
            params={'dag_id': TEST_DAG_ID, 'task_id': TEST_TASK_ID},
```

argibbs replied:
Ditto "I-blame-black"

josh-fell commented on tests/sensors/test_external_task_sensor.py:
```diff
@@ -134,17 +134,36 @@ def test_external_task_sensor_failed_states_as_success(self):
-                'dag_id': TEST_DAG_ID,
-                'task_id': TEST_TASK_ID,
-            },
+            params={'dag_id': TEST_DAG_ID, 'task_id': TEST_TASK_ID,},
```
```suggestion
            params={'dag_id': TEST_DAG_ID, 'task_id': TEST_TASK_ID},
```

argibbs replied:
This is a black thing ... I don't mind removing, but just to warn you, if someone re-runs black on the file, it'll come back

josh-fell replied:
Of course. Maybe CI black likes the one line, maybe it doesn't. Let's see! Totally fine if it comes back. If black is fine with the one line, I didn't want you to get dinged for it.

josh-fell commented on tests/sensors/test_external_task_sensor.py:
```diff
@@ -214,6 +230,31 @@ def test_external_dag_sensor(self):
+            external_dag_id='other_dag',
+            external_task_id=None,
+            allowed_states=["failed"],
+            failed_states=["success"],
```
```suggestion
            allowed_states=[State.Failed],
            failed_states=[State.Success],
```
WDYT about using the class attr rather than the harcoded value?

argibbs replied:
Good call.

josh-fell reviewed (commented):
This is very cool! I love all the extra documentation in the operator too; very helpful.  Some nits and suggestions so other static checks are happier before releasing all of the other ones in CI. And it's a good idea to [run the static checks locally](https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst) too if you can.

argibbs committed:
Apply suggestions from code review

- Removing some black-induced-trailing-commas
- Using explicit enum values rather than strings for state
- Joining two strings black pulled up onto the same line

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>

argibbs committed:
Applying black changes

argibbs committed:
Fixing tests

josh-fell reviewed (approved):
Pending "fun with tests". Great addition!
